addons/resource/models/resource.py

```diff
@@ -944,16 +944,16 @@ def _adjust_to_calendar(self, start, end):
         end, revert_end_tz = make_aware(end)
         result = {}
         for resource in self:
-            calendar_start = resource.calendar_id._get_closest_work_time(start, resource=resource)
-            search_range = None
             tz = timezone(resource.tz)
-            if calendar_start and start.astimezone(tz).date() == end.astimezone(tz).date():
+            calendar_start = resource.calendar_id._get_closest_work_time(start.replace(tzinfo=tz), resource=resource)
+            search_range = None
+            if calendar_start and start.astimezone(start.tzinfo).date() == end.astimezone(end.tzinfo).date():
                 # Make sure to only search end after start
                 search_range = (
-                    start,
-                    end + relativedelta(days=1, hour=0, minute=0, second=0),
+                    start.replace(tzinfo=tz),
+                    end.replace(tzinfo=tz) + relativedelta(days=1, hour=0, minute=0, second=0),
                 )
-            calendar_end = resource.calendar_id._get_closest_work_time(end, match_end=True, resource=resource, search_range=search_range)
```

Comment:
No search range is passed to _get_closest_work_time, the method will return the closest shift to end after end is set at the resource timezone.
`end: day, 21:59:59 + 5:30 => day+1, 2:29:59`
The closest shift will be the morning shift at day + 1, with match_end=True, _get_closest_work_time will return the end of the morning shift: 12:00